### DISCUSSION TITLE
Multiline JSON doesn't multiline

### DISCUSSION THREAD

**Discussion by henke443** [json5/json5#328] [Q&A] [answered]
I'm not sure if this is a bug or a feature request.

I want to use JSON5 as an alternative to YAML in an extension for VSCode.

In YAML really long strings are automatically multilined when stringified.

This doesn't seem to be the case for JSON5.

It doesn't stop here though. It seems that no matter how much I try (several hours later) I'm unable to automatically wrap 100+ character lines in strings. I've tried every possible combination of regex replace and reviver and replacer functions in parse and stringify, and it doesn't happen.

I think I will have to resort to using a regex replace that sees if it's within quotes or single quotes and then does the word wrap replace, but this is difficult and annoying. 

Can you add an example on how to do this in the README as this is something that probably a good chunk of users of this package would need to do. 

It would also be nice to have clear documentation of how the stringify replacer and the parse reviver works. Are newlines transformed before or after these methods are run?

Nice format and package otherwise.

**Comment by jordanbtucker** [comment, marked as answer]
Thanks for your interest in JSON5. However, the reason you're having problems with multiline strings is because JSON5 doesn't support multiline strings. It does support multiline comments, though.

As far as the documentation for `reviver` and `replacer`, the API is identical to `JSON`'s and works in the same way. [`reviver`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse#the_reviver_parameter) is called *after* each value is parsed while [`replacer`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#the_replacer_parameter) is called *before* each value is stringified.

**Reply by henke443** [reply]
Thanks, but you do support backslash multiline strings though, right?

The issue I was having is that the new lines I added with the replacer kept getting escaped by the stringify function, or atleast so I think. 

I will investigate more tomorrow when I'm more rested and more zen, really need to sleep now.

If it is the case that stringify is called before the value is stringified--and that this means new line escaping and such happens after the replacer has done changes--would it l be possible and make sense to add an option to replace the values after they are stringified instead?

**Reply by jordanbtucker** [reply]
I don't think I had my coffee this morning. I wasn't thinking about escaped line breaks (called line continuations) as multiline strings since ES6 has template syntax that does support line breaks inside strings. But you're right, this is a kind of multiline string in a way.

JSON5 (and ES5) does allow you to break your strings across lines by escaping line breaks. However, the behavior is slightly different from YAML's block scalars.

In YAML, block scalars can either preserve line breaks or replace them with spaces. In JSON5, line continuations will neither be replaced with spaces nor will they be replaced with line breaks. Instead, line continuations are treated as if they did not exist in the first place.

```json5
'Line 1\
Line 2'
// Equivalent to 'Line 1Line 2'
```

There is no way to represent a line break *as-is* in a string in JSON5. But you can do this:

```json5
'Line 1\n\
Line 2'
// Equivalent to 'Line 1\nLine 2'
```

As of now, this reference implementation of JSON5 does not have an option to include line continuations when stringifying, nor does it have a way to report whether and where line continuations exist in the source text when parsing (although this may change with #185).

In the interest of keeping the reference implementation lean, I don't plan on adding a feature for stringifying line continuations, but there may be [other implementations](https://github.com/json5/json5/wiki/In-the-Wild) that do support this kind of feature.

**Comment by jordanbtucker** [comment]
The reason that I moved this to a discussion is because the original issue doesn't work as a feature request either. JSON5 is built on the syntax of ECMAScript 5, which doesn't support multiline strings. So JSON5 will not have added support for multiline strings either.
